File: 2016/04/rs/src/main.rs
use regex::Regex;
use std::collections::HashMap;
use std::env;
use std::fs;
use std::time::Instant;

type Room = (String, u32, String);

fn is_room_valid(name: &str, checksum: &str) -> bool {
    let name = name.replace("-", "");
    let counts: HashMap<char, usize> = name.chars().map(|c| (c, name.matches(c).count())).collect();
    let mut counts: Vec<(&char, &usize)> = counts.iter().collect::<Vec<(&char, &usize)>>();
    counts.sort_by(|a, b| b.0.cmp(&a.0));
    counts.sort_by(|a, b| a.1.cmp(&b.1));
    counts.reverse();
    return checksum == counts.iter().map(|(c, _)| *c).take(5).collect::<String>();
}

const A_ORD: u8 = 97;
const Z_ORD: u8 = 122;
const DASH_ORD: u8 = 45;
const SPACE_ORD: u8 = 32;
fn get_next_char(c: u8) -> u8 {
    match c {
        DASH_ORD | SPACE_ORD => SPACE_ORD,
        Z_ORD => A_ORD,
        _ => c + 1,
    }
}

fn rotate_name(name: &str, count: u32) -> String {
    let mut name_ints = name.chars().map(|c| c as u8).collect::<Vec<u8>>();
    for _ in 0..count {
        for index in 0..name_ints.len() {
            name_ints[index] = get_next_char(name_ints[index]);
        }
    }
    name_ints.iter().map(|c| *c as char).collect::<String>()
}

const SEARCH_NAME: &str = "northpole object storage";
fn solve(rooms: &Vec<Room>) -> (u32, u32) {
    (
        rooms
            .iter()
            .filter(|(name, _, checksum)| is_room_valid(name, checksum))
            .map(|room| room.1)
            .sum(),
        rooms
            .iter()
            .filter(|(name, id, checksum)| {
                is_room_valid(name, checksum) && rotate_name(name, *id) == SEARCH_NAME
            })
            .next()
            .expect("Room not found")
            .1,
    )
}

fn get_input(file_path: &String) -> Vec<Room> {
    let re = Regex::new(r"^(?P<name>[a-z\-]+)-(?P<id>\d+)\[(?P<checksum>\w+)\]$").unwrap();
    fs::read_to_string(file_path)
        .expect("Error reading input file!")
        .lines()
        .map(|line| {
            re.captures(line)
                .map(|cap| {
                    (
                        String::from(cap.name("name").map(|c| c.as_str()).unwrap()),
                        cap.name("id").map(|c| c.as_str().parse().unwrap()).unwrap(),
                        String::from(cap.name("checksum").map(|c| c.as_str()).unwrap()),
                    )
                })
                .unwrap()
        })
        .collect()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:7}", now.elapsed().as_secs_f32());
}

File: 2016/04/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <regex.h>
#include <string.h>

typedef struct
{
    char *name;
    int id;
    char *checksum;
} Room;

typedef struct
{
    Room *rooms;
    int count;
    int size;
} Input;
typedef struct
{
    int part1;
    int part2;
} Results;

int isRoomValid(const char *name, const char *checksum)
{
    int processedCount = 0, letterIndex, processedIndex, shiftFrom, previousValue, oldValue;
    int counts[26];
    for (letterIndex = 0; letterIndex < 26; letterIndex++)
        counts[letterIndex] = 0;
    char c;
    while ((c = *(name++)))
        if (c >= 'a')
            counts[c - 'a']++;
    char processed[5];
    for (letterIndex = 0; letterIndex < 26; letterIndex++)
    {
        shiftFrom = -1;
        for (processedIndex = 0; processedIndex < processedCount; processedIndex++)
            if (counts[letterIndex] > counts[processed[processedIndex]] ||
                (counts[letterIndex] == counts[processed[processedIndex]] && letterIndex < processed[processedIndex]))
            {
                shiftFrom = processedIndex;
                break;
            }
        if (shiftFrom != -1)
        {
            processedCount++;
            previousValue = letterIndex;
            for (; processedIndex < processedCount && processedIndex < 5; processedIndex++)
            {
                oldValue = processed[processedIndex];
                processed[processedIndex] = previousValue;
                previousValue = oldValue;
            }
        }
        else if (processedCount < 5)
            processed[processedCount++] = letterIndex;
    }
    for (int processedIndex = 0; processedIndex < 5; processedIndex++)
        processed[processedIndex] += 'a';
    return strncmp(processed, checksum, 5) == 0;
}

int part1(Input input)
{
    int total = 0;
    while (input.count--)
    {
        if (isRoomValid(input.rooms->name, input.rooms->checksum))
            total += input.rooms->id;
        input.rooms++;
    }
    return total;
}

char getNextChar(char c)
{
    switch (c)
    {
    case '-':
    case ' ':
        return ' ';
    case 'z':
        return 'a';
    default:
        return c + 1;
    }
}

void rotateName(char *name, int count)
{
    while (count--)
    {
        int index = 0;
        char c;
        while ((c = name[index]) != 0)
            name[index++] = getNextChar(c);
    }
}

const char *SEARCH_NAME = "northpole object storage";
int part2(Input input)
{
    while (input.count--)
    {
        if (isRoomValid(input.rooms->name, input.rooms->checksum))
        {
            rotateName(input.rooms->name, input.rooms->id);
            if (strcmp(SEARCH_NAME, input.rooms->name) == 0)
                return input.rooms->id;
        }
        input.rooms++;
    }
    return -1;
}

Results solve(Input input)
{
    return (Results){part1(input), part2(input)};
}

#define INPUT_INCREMENT 10
void addToInput(Input *input, char *name, int id, char *checksum)
{
    if (input->count == input->size)
    {
        input->size += INPUT_INCREMENT;
        input->rooms = realloc(input->rooms, input->size * sizeof(Room));
    }
    input->rooms[input->count++] = (Room){name, id, checksum};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    regex_t regexCompiled;
    regmatch_t groupArray[4];
    if (regcomp(&regexCompiled, "^([a-z\\-]+)-([0-9]+)\\[([a-z]+)\\]", REG_EXTENDED))
    {
        perror("Error compiling regex.");
        exit(1);
    }
    size_t lineLength;
    char *line = NULL, *cursor = NULL;
    Input input = {
        calloc(INPUT_INCREMENT, sizeof(Room)),
        0,
        INPUT_INCREMENT};
    char *name, *checksum;
    int id, group, groupLength;
    while (getline(&line, &lineLength, file) != EOF)
    {
        cursor = line;
        while (!regexec(&regexCompiled, cursor, 4, groupArray, 0))
        {
            for (group = 0; group <= 3 && groupArray[group].rm_so != -1; group++)
            {
                groupLength = strlen(cursor) + 1;
                char cursorCopy[groupLength];
                strcpy(cursorCopy, cursor);
                cursorCopy[groupArray[group].rm_eo] = 0;
                switch (group)
                {
                case 1:
                    name = malloc(groupLength);
                    strcpy(name, cursorCopy + groupArray[group].rm_so);
                    break;
                case 2:
                    id = atoi(cursorCopy + groupArray[group].rm_so);
                    break;
                case 3:
                    checksum = malloc(groupLength);
                    strcpy(checksum, cursorCopy + groupArray[group].rm_so);
                }
            }
            cursor += groupArray[0].rm_eo;
        }
        addToInput(&input, name, id, checksum);
    }
    fclose(file);
    return input;
}

void freeInput(Input input)
{
    Room *rooms = input.rooms;
    while (input.count--)
    {
        free(rooms->name);
        free(rooms->checksum);
        rooms++;
    }
    free(input.rooms);
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
